cranelift-codegen/src/peephole.rs

```diff
@@ -0,0 +1,469 @@
+                if si.offs == offs && si.size == size {
+                    if reg2 == reg {
+                        panic!("has_inexact_binding: binding *is* exact!");
+                    }
+                    return Some((reg2, si.value));
+                }
+            }
+        }
+        // No such binding.
+        None
+    }
+
+    // Create the binding `(reg, value, offs, size)` in `env`, and throw away
+    // any previous binding associated with `reg`.
+    fn bind(&mut self, reg: RegUnit, value: Value, offs: i32, size: u32) {
+        debug_assert!(self.map.len() == self.size);
+        self.map[reg as usize] = Some(SlotInfo { value, offs, size });
+    }
+}
+
+impl<'a> Context<'a> {
+    /// Run the post-regalloc peephole optimiser.
+    fn cleanup(&mut self) {
+        let mut tot_regunits = 0;
+        for ru in self.reginfo.banks {
+            tot_regunits += ru.units;
+        }
+        // Perform redundant-fill removal.  We can visit the EBBs in any
+        // order, since there's no context carried across EBB boundaries at
+        // present.
+        for &ebb in self.domtree.cfg_postorder().iter().rev() {
+            self.cleanup_ebb(ebb, tot_regunits);
+        }
+    }
+
+    // Returns the regunit that `v` is allocated to.  Assumes that `v`
+    // actually resides in a register.
+    fn reg_of_value(&self, v: Value) -> RegUnit {
+        match self.cur.func.locations[v] {
+            ValueLoc::Reg(ru) => ru,
+            _ => panic!("reg_of_value: value isn't in a reg"),
+        }
+    }
+
+    // Returns the stack slot that `v` is allocated to.  Assumes that `v`
+    // actually resides in a stack slot.
+    fn slot_of_value(&self, v: Value) -> &StackSlotData {
+        match self.cur.func.locations[v] {
+            ValueLoc::Stack(slot) => &self.cur.func.stack_slots[slot],
+            _ => panic!("slot_of_value: value isn't in a stack slot"),
+        }
+    }
+
+    // Returns the regunits written by `inst`.  May include duplicates.
+    fn get_regs_w(&self, inst: Inst) -> Vec<RegUnit> {
+        let dst_vals: &[Value] = self.cur.func.dfg.inst_results(inst);
+        let mut dst_regs = Vec::<RegUnit>::new();
+        for v in dst_vals.iter() {
+            if let ValueLoc::Reg(ru) = self.cur.func.locations[*v] {
+                dst_regs.push(ru);
+            }
+        }
+        dst_regs
+    }
+
+    // This is the main EBB processing function.  It carries an `AvailEnv`
+    // forwards along the block, using it to detect redundant fills.
+    fn cleanup_ebb(&mut self, ebb: Ebb, tot_regunits: u16) {
+        // A mapping from [0 .. tot_regunits) to `Option<SlotInfo>`
+        let mut env = AvailEnv::new(tot_regunits as usize);
+
+        self.cur.goto_top(ebb);
+        while let Some(inst) = self.cur.next_inst() {
+            debug_assert!(env.invariants_ok(), "cleanup_ebb: env invariants not ok");
+
+            // To avoid difficulties with the borrow checker, do this in two
```

Comment:
For the record, which difficulties are you avoiding here? (I'm wondering if, with little extra effort, we could avoid the indirection and cognitive load that `replacement` materializes.)

Reply:
Borrowing-mutable something that is already borrowed-immutable, is my guess -- I seem to run into this a lot when trying to iterate over the instructions to update them, and have ended up with the same structure, an information-gathering pass and an updating pass.

Reply:
@bnjbvr: yes, it's what Lars says.  Regarding cognitive load, I sometimes find it easier to read the style where we first gather information, then transform, than to understand combined analysis and transformation logic -- although that's not the motivating factor here.